lib/bolt_server/pe/pal.rb

```diff
@@ -64,17 +67,28 @@ def initialize(plan_executor_config, environment_name, hiera_config = nil, max_c
           # environment: _technically_ that environment always exists, but if the dir
           # isn't there it won't find the module and fail with "plan not found" rather
           # than "environment doesn't exist"
-          if environment.configuration.path_to_env
-            bolt_config = File.join(environment.configuration.path_to_env, 'bolt.yaml')
+          if path_to_env
+            bolt_yaml = File.join(environment.configuration.path_to_env, 'bolt.yaml')
+            modulepath_setting_from_bolt = Bolt::Util.read_optional_yaml_hash(bolt_yaml, 'config')['modulepath']
+          end
+
+          # If we loaded a bolt.yaml in the environment root and it contained a modulepath setting:
+          # we will use that modulepath rather than the one loaded through puppet. modulepath will
+          # be the _only_ setting that will work from bolt.yaml in plans in PE.
+          if modulepath_setting_from_bolt
+            modulepath_setting_from_bolt.split(File::PATH_SEPARATOR).each do |path|
+              if Pathname.new(path).absolute? && File.exist?(path)
+                modulepath_dirs << path
+              elsif File.exist?(File.join(path_to_env, path))
+                modulepath_dirs << File.join(path_to_env, path)
+              end
+            end
```

Comment:
What if this ends up being an empty array? Say, none of the module paths in the bolt.yaml exist.

Reply:
`modulepath_settings_from_bolt` will be `nil` and `modulepath_dirs` will be `environment.modulepath`

Reply:
I think that just the modules on the basemodulepath will be available.        https://github.com/puppetlabs/bolt/pull/1572/commits/93f7fc6472192ce7e1ceae9ad416a32867a362a4#diff-bef67b4dd1c8e5fa42a6a70e7cd5d182R88

Reply:
Yeah, that's what it looks like - is that desirable? Maybe we should log some warnings or something for the missing paths? This seems like it would be difficult to debug if something goes wrong without some sort of logging.

Reply:
Using bolt-server to list plans is kind of strange given that it does not run plans. I removed all logging in BoltServer::PE::PAL bc i thought logging plan stuff to bolt-server logs would be confusing.